test/libsolidity/SolidityEndToEndTest.cpp

```diff
@@ -11901,6 +11901,566 @@ BOOST_AUTO_TEST_CASE(senders_balance)
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(createFixed(225, 100, 18)));
+}
+
+BOOST_AUTO_TEST_CASE(fixed_subtraction)
+{
+	char const* sourceCode = R"(
+		contract test {
+			fixed a = 3.5;
+			fixed b = 1.25;
+
+			function f() returns (fixed) {
+				return b - a;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(createFixed(-225, 100, 18)));
+}
+
+BOOST_AUTO_TEST_CASE(ufixed_multiplication)
+{
+	char const* sourceCode = R"(
+		contract test {
+			ufixed a = 3.5;
+			ufixed b = 1.25;
+
+			function f() returns (ufixed) {
+				return a * b;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(createFixed(4375, 1000, 18)));
+}
+
+BOOST_AUTO_TEST_CASE(fixed_multiplication)
+{
+	char const* sourceCode = R"(
+		contract test {
+			fixed a = 3.5;
+			fixed b = -1.25;
+
+			function f() returns (fixed) {
+				return a * b;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(createFixed(-4375, 1000, 18)));
+}
+
+BOOST_AUTO_TEST_CASE(ufixed_division_large_divisor)
+{
+	char const* sourceCode = R"(
+		contract test {
+			ufixed a = 3.5;
+			ufixed b = 1.25;
+
+			function f() returns (ufixed) {
+				return a / b;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(createFixed(28, 10, 18)));
+}
+
+BOOST_AUTO_TEST_CASE(fixed_division_large_divisor)
+{
+	char const* sourceCode = R"(
+		contract test {
+			fixed a = 3.5;
+			fixed b = -1.25;
+
+			function f() returns (fixed) {
+				return a / b;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(createFixed(-28, 10, 18)));
+}
+
+BOOST_AUTO_TEST_CASE(ufixed_division_small_divisor)
+{
+	char const* sourceCode = R"(
+		contract test {
+			ufixed a = 3.5;
+			ufixed b = 0.125;
+
+			function f() returns (ufixed) {
+				return a / b;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(createFixed(28, 1, 18)));
+}
+
+BOOST_AUTO_TEST_CASE(fixed_division_small_divisor)
+{
+	char const* sourceCode = R"(
+		contract test {
+			fixed a = 3.5;
+			fixed b = -0.125;
+
+			function f() returns (fixed) {
+				return a / b;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(createFixed(-28, 1, 18)));
+}
+
+BOOST_AUTO_TEST_CASE(ufixed_modulo)
+{
+	char const* sourceCode = R"(
+		contract test {
+			ufixed a = 5.7;
+			ufixed b = 2.5;
+
+			function f() returns (ufixed) {
+				return a % b;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(createFixed(7, 10, 18)));
+}
+
+BOOST_AUTO_TEST_CASE(fixed_modulo)
+{
+	char const* sourceCode = R"(
+		contract test {
+			fixed a = 5.7;
+			fixed b = -2.5;
+
+			function f() returns (fixed) {
+				return a % b;
```

Comment:
Why should this support negative modulo?

Reply:
Negative modulus is well-defined.